contracts/ArchetypeLogic.sol

```diff
@@ -0,0 +1,319 @@
+  uint16 affiliateDiscount; //BPS
+  MintTier[] mintTiers;
+}
+
+struct Config {
+  string baseUri;
+  address affiliateSigner;
+  address ownerAltPayout; // optional alternative address for owner withdrawals.
+  address superAffiliatePayout; // optional super affiliate address, will receive half of platform fee if set.
+  uint32 maxSupply;
+  uint32 maxBatchSize;
+  uint16 affiliateFee; //BPS
+  uint16 platformFee; //BPS
+  uint16 defaultRoyalty; //BPS
+  Discount discounts;
+}
+
+struct Options {
+  bool uriLocked;
+  bool maxSupplyLocked;
+  bool affiliateFeeLocked;
+  bool discountsLocked;
+  bool ownerAltPayoutLocked;
+  bool royaltyEnforcementEnabled;
+  bool royaltyEnforcementLocked;
+  bool provenanceHashLocked;
+}
+
+struct DutchInvite {
+  uint128 price;
+  uint128 reservePrice;
+  uint128 delta;
+  uint32 start;
+  uint32 limit;
+  uint32 maxSupply;
+  uint32 interval;
+  address tokenAddress;
+}
+
+struct Invite {
+  uint128 price;
+  uint32 start;
+  uint32 limit;
+  uint32 maxSupply;
+  address tokenAddress;
+}
+
+struct OwnerBalance {
+  uint128 owner;
+  uint128 platform;
+}
+
+struct BurnConfig {
+  IERC721AUpgradeable archetype;
+  bool enabled;
+  uint16 ratio;
+  uint64 start;
+  uint64 limit;
+}
+
+address constant PLATFORM = 0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC; // TEST (account[2])
+// address private constant PLATFORM = 0x86B82972282Dd22348374bC63fd21620F7ED847B;
+uint16 constant MAXBPS = 5000; // max fee or discount is 50%
+
+library ArchetypeLogic {
+  // calculate price based on affiliate usage and mint discounts
+  function computePrice(
+    DutchInvite storage invite,
+    Discount storage discounts,
+    uint256 numTokens,
+    bool affiliateUsed
+  ) public view returns (uint256) {
+    uint256 price = invite.price;
+    if (invite.interval != 0) {
+      uint256 diff = (((block.timestamp - invite.start) / invite.interval) * invite.delta);
+      if (price > invite.reservePrice) {
+        if (diff > price - invite.reservePrice) {
+          price = invite.reservePrice;
+        } else {
+          price = price - diff;
+        }
+      } else if (price < invite.reservePrice) {
+        if (diff > invite.reservePrice - price) {
+          price = invite.reservePrice;
+        } else {
+          price = price + diff;
+        }
+      }
+    }
+
+    uint256 cost = price * numTokens;
+
+    if (affiliateUsed) {
+      cost = cost - ((cost * discounts.affiliateDiscount) / 10000);
+    }
+
+    for (uint256 i = 0; i < discounts.mintTiers.length; i++) {
+      if (numTokens >= discounts.mintTiers[i].numMints) {
+        return cost = cost - ((cost * discounts.mintTiers[i].mintDiscount) / 10000);
+      }
+    }
+    return cost;
+  }
+
+  function validateMint(
+    DutchInvite storage i,
+    Config storage config,
+    Auth calldata auth,
+    uint256 quantity,
+    address owner,
+    address affiliate,
+    uint256 curSupply,
+    mapping(address => mapping(bytes32 => uint256)) storage minted,
```

Comment:
The storage keyword lets state variables be passed by reference, ends up saving gas since less state reads.